### ISSUE TITLE
Support `when`-like condition

### ISSUE THREAD

**Issue by stevenguh** [VSpaceCode/vscode-which-key#11]
Split from #5.

My current thinking on this is at https://github.com/VSpaceCode/vscode-which-key/issues/5#issuecomment-667554135

**Comment by stevenguh** [comment]
Just jotting here. We will likely need another context variable like `whichKeyVisible` so we the the shortcut hack won't be executing when if the command is long running like renaming a reference/variable.

**Comment by MarcoIeni** [comment]
could this enable major mode?

**Comment by stevenguh** [comment]
Can you elaborate a little bit more? I am not familiar with major mode.

**Comment by MarcoIeni** [comment]
In spacemacs you have major-modes, like "python", "markdown", and so on. So for example if you open a `.md` file spacemacs will use Markdown mode in that buffer.

When you are using a mode, key bindings under `<spc> m` or under `,` will be unique to that mode.

This means that when you are editing a markdown, `<spc> m b` could be assigned to "make bold", while in python `<spc> m b` could be assigned to "toggle breakpoint" (I am just inventing).

**Comment by stevenguh** [comment]
Thank you for the clear explanation. I think this work should take this into account and facilitate major mode as well if we are implementing a type called "conditional".

**Comment by stevenguh** [comment]
I have a working prototype on `conditionals` branch; however, as I am trying implement overrides (add/delete/modify) for conditional bindings. I realized that the `condition` in a conditional binding is basically the `key` in regular binding but instead of string, we defined it as object.

If a user wants to redefined the condition, the user can use the following example to override the whole conditional binding.
```json
{
    "bindingOverrides": [
        {
            "keys": "m",
            "name": "Major",
            "conditional": [
                {
                    "type": "bindings",
                    "bindings": []
                },
                {
                    "type": "bindings",
                    "name": "JavaScript",
                    "bindings": [],
                    "condition": {
                        "langId": "javascript"
                    }
                }
            ]
        }
    ]
}
```

In order to implement the traversal of the conditional bindings, I have to re-use the keys in the current overrides format in the following example.
The example is an override for conditional binding with a key of `m`. In this case the last key will be read as condition for the command with the format of a query string. This will add/modify the conditionals depending if "languageId=typescript" exists. This is also powerful that it can traverse through the conditional bindings (examples keys: `["m", "languageId=typescript", "i"]`). Note, this is possible because we do not implicitly any bindings if the part of the key sequence doesn't exist in the overrides.
```json
"bindingOverrides": [
    {
        "keys": ["m", "languageId=typescript"],
        "name": "123",
        "type": "command",
        "command": "workbench.action.editor.changeLanguageMode"
    }
]
```

Although, this works for the most part, There are two things that bug me with the current implementation.
1. **The asymmetry of bindings and overrides** (one using object to present condition and other use a query string)
Working the issue backward, in order to unify them I think we can use query string to represent the condition as key, and then we can reuse the "bindings" key without introducing a new key. An example binding is as follow:
```jsonc
{
    "key": "m",
    "name": "Major mode",
    "type": "conditional",
    "bindings": [
        {
            "key": "",
            "type": "bindings",
            "bindings": [...],
        },
        {
            "key": "langId=javascript"
            "type": "bindings",
            "bindings": [...],
        }
    ],
}
```
2. **Needs to escape query string for some character**
This is especially prominent if we have a "when" condition that uses `&` (e.g. `sideBarVisible && explorerViewletVisible`) and it makes the config not humanly readable. The example override with escape characters.
```json
"bindingOverrides": [
    {
        "keys": [
            "m",
            "when=sideBarVisible%20%26%26%20explorerViewletVisible"
        ],
        "name": "123",
        "type": "command",
        "command": "workbench.action.editor.changeLanguageMode"
    }
]
```
---
I am thinking
- Use a custom query string that uses `;` as separator instead of `&`
- Use a custom query string as the `key` for conditional binding instead of `condition` like I mentioned above.
- Instead of adding a new key `conditionals`, we can then reuse the key `bindings`
- This has also the added benefit of reuse the same schema validation
